### ISSUE TITLE
Custom comparator ignored for comparisons to self

### ISSUE THREAD

**Issue by mknejp** [nunit/nunit#3692]
The following test succeeds:
```cs
public static void Test()
{
    var a = new object();

    Assert.That(a, Is.EqualTo(a).Using<object>((object left, object right) => false));
}
```
However, I believe it should fail. If I provide a custom comparator it should always take precedence. The way NUnit behaves right now means I cannot, for example, give it a custom comparator that ensures user-defined `a == b` or `a != b` operators handle self-tests correctly.

Tested with NUnit 4.12.0

**Comment by mikkelbu** [comment]
I agree that this sounds like an error. The current comparison perform null checks and `ReferenceEquals` before determining whether it should use an externalComparer (aka custom comparer). Changing the existing behaviour could break peoples' tests (e.g. if they assume that the above checks have already been performed, but to me it feels like they - in that case - are depending on an implementation detail).

https://github.com/nunit/nunit/blob/85432008183bfcbd44994726f14525b2ea897fb7/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs#L170-L188

**Comment by mknejp** [comment]
> Changing the existing behaviour could break peoples' tests

Which is probably a good thing, because they currently aren't testing their custom comparator for these cases (but probably should), possibly masking bugs in those.

**Comment by manfred-brands** [comment]
What would be a practical situation where an object doesn't compare Equal with itself? It would be impossible to find such an object in an Dictionary. It also violates the first rule in [Object.Equals](https://docs.microsoft.com/en-us/dotnet/api/system.object.equals?view=net-5.0#System_Object_Equals_System_Object_): `x.Equals(x) returns true`.

**Comment by mknejp** [comment]
That isn't the issue. The issue is that NUnit skips the custom comparator in those situations, which it shouldn't. You want to make sure (i.e. test) that the custom comparator handles `null` and self-tests correctly, which you currently can't using the `EqualTo().Using()` constraint.

**Comment by manfred-brands** [comment]
I see your point now.

In that case, it applies to everything, it should pass `null` and self always to any custom or normal `Equals` overloads.

The following bad `Equals` implementation is not detected:
```
[Test]
public void TestEquals()
{
    var instance = new ClassWithBadEquals("Boom");

    Assert.That(instance, Is.Not.EqualTo(null)); // Should have failed!
    Assert.That(() => instance.Equals(null), Throws.TypeOf<NullReferenceException>());
}

private class ClassWithBadEquals : IEquatable<ClassWithBadEquals>
{
    public ClassWithBadEquals(string message) => Message = message;

    public string Message { get; }

    public bool Equals(ClassWithBadEquals other) => other.Message == Message;

    public override bool Equals(object obj) => Equals(obj as ClassWithBadEquals);

    public override int GetHashCode() => 42;
}
```

If I wanted to explicitly test my `Equals` override : `Assert.That(instance.Equals(null), Is.False);`
This fires NUnit2010 with a code fix that turns it into: `Assert.That(instance, Is.Not.EqualTo(null));`

Which no longer even touches my `Equals` method, but test the NUnit one.

**Comment by manfred-brands** [comment]
The question then becomes: Should the behaviour of NUnit be changed or should we in the one case where we do actually test the implementation add suppression for rule NUnit2010? In all other cases the behaviour of our method is supposed to match the first lines highlighted by @mikkelbu .

**Comment by mknejp** [comment]
I'm on the "change it" side. If you want a hard test against `null` use `SameAs()` or `Is.[Not.]Null`. It should be clear that it bypasses any custom `Equals()` implementation. but if I know my type has a custom `Equals()` override and I write `Is.EqualTo(null)` Tests expecting it to be called, but it never does... that's quite a surprise (it was for me). I *definitely* expect a custom `Using(...)` comparator to be called with the original arguments in all cases. Otherwise the comparator isn't being fully exercised.

But I'm merely one guy stumbling upon this. It seems like this hasn't bothered anyone before. I wonder if my expectations are off  or if this is something people just haven't realized or thought about.

**Comment by stevenaw** [comment]
I agree with @mikkelbu that this can seem a little misleading, as well as that changing this could break others' code. As far as the the change itself, breaking changes are always a little riskier as it could be disruptive to others. Putting the "if we should" aside, and instead focusing on "when", NUnit follows semantic versioning for its releases and breaking changes would usually be reserved for major release like version NUnit 4.0.

@mknejp You may already know this, but if you're trying to test a custom comparer class and want to specifically test null/self scenarios, another option would be to call the comparer directly. A custom failure message may be desired with this approach to give informative reporting (otherwise you may just see "Expected: true but was false")

```csharp
class MyComparer : IEqualityComparer {
    public bool Equals(object x, object y) => false;
    public int GetHashCode() => 0;
}

[Test]
public static void Test()
{
    var a = new object();
    var comparer = new MyComparer();

    Assert.That(comparer.Equals(a, a), Is.True, $"{a} should be equal to {a}");
}
```

**Comment by mknejp** [comment]
I am not a great fan of not breaking code for the sake of not breaking code. The disruption of having to fix your `Equals()` implementations is low in comparison to the disruption of your `Equals()` crashing in production.

Anyone who currently expects `Assert.That(x, Is.EqualTo(y))` to _always_ invoke the corresponding `Equals()` override is basing their test results on a broken premise. If the author of the test expects `Assert.That(x, Is.EqualTo(null))` to actually test that `x.Equals(null)` is handled gracefully, then the test is meaningless _and_ misleading. For reference comparisons we have `Is.SameAs()`.

The [documentation](https://docs.nunit.org/articles/nunit/writing-tests/constraints/EqualConstraint.html) only mentions the custom comparator case, but says nothing about not calling `Equals()`. IMO clause 7 at the end of the page is a mistake, but it is documented behavior, so fine.

But the there's this about custom comparators:
> If the default NUnit or .NET behavior for testing equality doesn't meet your needs, you can supply a comparer of your own through the Using modifier. 

Except that is again misleading. If I don't like the null short-circuiting behavior of the built-in `EqualTo()` constraint then providing a custom comparator with `Using` is of no use because it has the exact same behavior.

I still believe that not calling `Equals()` or the comparator is a mistake, and the call should only be skipped where it is impossible to determine a target for the method call. The testing framework is deciding what equality means instead of the author of the class or comparator.

**Comment by gruenedd** [comment]
> But I'm merely one guy stumbling upon this.

Seems like we are at least two now ;-)
#3842

**Comment by rjnilsson** [comment]
I think I've just been bitten by this. Add my vote to a fix.

In my case, the requirement for my custom comparer is throw an exception by default if one tries to compare an instance to itself.

It is a pretty specific case intended to be used for comparing entities, as in mutable objects, where comparing an entity to itself very likely is _not_ what the test writer intended to.

**Comment by stevenaw** [comment]
I am going to close #3842 as a duplicate of this one. The use case cited there by @gruenedd is:

> In my use case i compare two lists for equality except when an item is null which means "do not compare here, i dont care at this index, this is not part of the test" (so technically i compare null with non-null items; strings in my case).
Sure, i could rewrite the test by looping by myself, but i want to use all that nice reporting stuff which is implemented in the NUnit enumerable/list equality comparers (difference by length, type, extra items and so on...) and i also failed to grab/reuse parts of it (see below) due too privacy settings.

**Comment by stevenaw** [comment]
@nunit/framework-team I've labelled this as an "idea" for now, though it seems there's already been a lot of discussion around it and general agreement that this might be helpful. Should we promote it to "enhancement" ?

**Comment by OsirisTerje** [comment]
> Should we promote it to "enhancement" ?

It makes sense, so sounds good to me.

**Comment by mikkelbu** [comment]
I've added an attempt at a PR - I'm concerned about this breaking a lot of people as comparators now needs to be able to handle a nullable instance - see e.g. src/NUnitFramework/tests/Constraints/AnyOfConstraintTests.cs in the PR.

I'm wondering if there is an better approach as we cannot have 2 overloads that only differ on nullability

**Comment by manfred-brands** [comment]
> I'm wondering if there is an better approach as we cannot have 2 overloads that only differ on nullability

You could add a `AllowCompareNull` modifier.
Only new code that uses this need to be aware.

**Comment by gruenedd** [comment]
> > I'm wondering if there is an better approach as we cannot have 2 overloads that only differ on nullability
> You could add a `AllowCompareNull` modifier. Only new code that uses this need to be aware.

Regardless of any additional settings or configuration options, the comparator should be invoked by default, including for null values. Otherwise, the main point of this issue is largely lost, since it is specifically about unexpected behavior.

**Comment by stevenaw** [comment]
@gruenedd One consideration to that is that having this on by default would likely break others, meaning we would need to wait for a major version release to do that. New major versions can be in development for a year or two - a long time to wait for a new feature.

Allowing it to be opt-in -at least at first- could allow us to release it earlier in a minor version of the framework meaning that yourself and others could then use it earlier as well.

@mikkelbu @manfred-brands  You're both more involved with this issue than I am at the moment and I'm trying to avoid advocating for any particular approach here with this comment.

**Comment by mikkelbu** [comment]
@stevenaw Note that this is for V5 and has label `Breaking Change`, but I'm just trying to determine the smallest possible breaking change :)

Notes that I need to think about (as I have to leave my house soon)
* We have changed the ordering in `NUnitEqualityComparer.AreEqual`, so now the instances passed to `externalComparer.AreEqual(...)` can now be null (before they couldn't as we had the `null` check before calling the external comparer)
* So this entails that `EqualityAdapter.AreEqual(object x, object y)` should now be able to handle `null` for `x` and `y`
* There are 7 classes that override this method (all private/internal), but it seems that the only problem is `PredicateEqualityAdapter` 

```csharp
        public static EqualityAdapter For<TExpected, TActual>(Func<TExpected, TActual, bool> comparison)
        {
            return new PredicateEqualityAdapter<TExpected, TActual>(comparison);
        }

        internal sealed class PredicateEqualityAdapter<TActual, TExpected> : EqualityAdapter
        {
            private readonly Func<TActual, TExpected, bool> _comparison;

            /// <summary>
            /// Returns true if the two objects can be compared by this adapter.
            /// The base adapter cannot handle IEnumerables except for strings.
            /// </summary>
            public override bool CanCompare(object x, object y)
            {
                return TypeHelper.CanCast<TExpected>(x) && TypeHelper.CanCast<TActual>(y);
            }

            /// <summary>
            /// Compares two objects, returning true if they are equal
            /// </summary>
            public override bool AreEqual(object x, object y)
            {
                return _comparison.Invoke((TActual)y, (TExpected)x);
            }

            public PredicateEqualityAdapter(Func<TActual, TExpected, bool> comparison)
            {
                _comparison = comparison;
            }
        }
```

So when when `AreEqual(object x, object y)` is changed to `AreEqual(object? x, object? y)` then we cannot do the cast unless we also change `_comparison` and add `?` to the method and the casts (and then we will affect all usages of `For`). I guess the problem is that there is little context between `x` / `y` and `TActual` / `TExpected`

EDIT: Perhaps I can also just have (although I really dislike `!`
```
public override bool AreEqual(object? x, object? y)
{
    return _comparison.Invoke((TActual)y!, (TExpected)x!);
}
```
as `GetExternalComparer` ensures that we only inspects `ExternalComparer`s that can compare the two instances
